File: scripts/algorithms/C/Count Subtrees With Max Distance Between Cities/Count Subtrees With Max Distance Between Cities.py
# Runtime: 2577 ms (Top 35.48%) | Memory: 14.1 MB (Top 20.97%)
from collections import defaultdict
class Solution:
    def countSubgraphsForEachDiameter(self, n: int, edges: List[List[int]]) -> List[int]:
        def floydwarshall(n, edges):
            D = [[float('inf')]*n for _ in range(n)]
            for u, v in edges:
                D[u-1][v-1] = 1
                D[v-1][u-1] = 1
            for v in range(n):
                D[v][v] = 0
            for k in range(n):
                for i in range(n):
                    for j in range(n):
                        D[i][j] = min(D[i][j],D[i][k]+D[k][j])
            return D

        G = defaultdict(list)
        for v, w in edges:
            G[v-1].append(w-1)
            G[w-1].append(v-1)

        def dfs(G, v, V, visited):
            if v in visited: return
            visited.add(v)
            for w in G[v]:
                if w in V:
                    dfs(G, w, V, visited)

        def check_connected(G, V):
            if not V: return False
            root = next(iter(V))
            visited = set()
            dfs(G, root, V, visited)
            return visited == V

        def max_distance(D, V):
            res = 0
            for v in V:
                for w in V:
                    res = max(res, D[v][w])
            return res

        def solve(include, idx, ans, G, D):
            if idx == n:
                V = set(v for v in range(n) if include[v])
                if check_connected(G, V):
                    d = max_distance(D, V)
                    if d >= 1:
                        ans[d-1] += 1
            else:
                solve(include, idx+1, ans, G, D)
                include[idx] = True
                solve(include, idx+1, ans, G, D)
                include[idx] = False

        D = floydwarshall(n, edges)
        include = [False] * n
        ans = [0]*(n-1)
        solve(include, 0, ans, G, D)
        return ans

File: scripts/algorithms/C/Count Subtrees With Max Distance Between Cities/Count Subtrees With Max Distance Between Cities.java
// Runtime: 232 ms (Top 14.5%) | Memory: 44.30 MB (Top 38.1%)

class Solution {
    public int[] countSubgraphsForEachDiameter(int n, int[][] edges) {
         
        
       ArrayList<Integer>[] graph = (ArrayList<Integer>[]) new ArrayList[n];
       for (int i = 0; i < n; i++) graph[i] = new ArrayList<>();
        
         for(int i=0;i<edges.length;i++)
         {
             int u = edges[i][0];
             int v = edges[i][1];
             
             --u;
             --v;
             
             graph[u].add(v);
             graph[v].add(u);
         }
        
         int[] darr = new int[n-1];
        
         for(int mask=0;mask<(1<<(n));mask++)
         {
             int d = getmax(mask , graph);
             
             if(d>0)
             {
                 darr[d-1]++;
             }
         }
        
        return darr;
         
    }
    
   int getmax(int mask , ArrayList<Integer>[] graph)
    {
          
          int maxd = -1;
        
          for(int i=0;i<15;i++)
          {      
              if( (mask & (1<<i)) !=0)
              {
                  maxd = Math.max(maxd , dfs(i , graph , mask));
              }
          }
        
        return maxd;
    }
    
    int dfs(int node , ArrayList<Integer>[] graph , int mask)
    {
        
        
      Queue<Integer> q
            = new LinkedList<>();
        q.add(node);
        int dist = -1;
        int curr = mask;
        
        while(!q.isEmpty())
        {
           dist++;
        
         for(int l=q.size()-1;l>=0;l--) {   
            int z1 = q.peek();
            curr = curr ^ (1<<z1);
            q.remove();
            
            for(int z : graph[z1])
            {
               if(((mask & (1<<z))!=0) && ((curr & (1<<z))!=0))
               {  
                  q.add(z); }
            }
         }
                  
        }
        
        if(curr!=0)
        {
            return -1;
        }
        
        return dist;
       
    }
}